<a href="https://colab.research.google.com/github/AEIBIANCOFCE/AnalisisEstadisticoI/blob/main/7%20N%C3%BAmeros%20%C3%8Dndice/9_AEI_N%C3%BAmeros_%C3%8Dndice_Python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<figure>
<center>
<img src='https://www.economicas.uba.ar/wp-content/uploads/2020/08/cropped-logo_FCE.png' />
</figure>

# **Universidad de Buenos Aires**
## **Facultad de Ciencias Económicas**

### **Análisis Estadístico I**

### Cátedra: Bianco

#### **Números Índice con Python**

***Material elaborado por la profesora Natalia Salaberry***

####**Librerías**

In [ ]:
import numpy as np

import yfinance as yf
import pandas as pd

####**Datos**

Se obtienen datos de la web [YahooFinance](https://finance.yahoo.com/?guccounter=1&guce_referrer=aHR0cHM6Ly93d3cuZ29vZ2xlLmNvbS8&guce_referrer_sig=AQAAANeM5yh3YzGRXBQrt0GywmqcBh-DHbrvwUShNKENrv-8D-M5ZRwYuV8Qq3CZaPD1HGC71X3WATrcHImLnNUTP2U0C9wRJUjNoOlu6beTkb_xbMpvtldSESLYt67gir69my9vOcC02T1Bq3P4xD7suDQeQW4-A6vu1rm8gV5HoRvB) a través de conexión a Api (mediante librería) sobre el activo Google

In [ ]:
pd.set_option("display.precision", 14) #visualizar cantidad de decimales completa
GOOG=pd.DataFrame(yf.download('GOOG', start='2024-08-01', end='2024-09-01')[['Close', 'Volume']]).reset_index()
GOOG

[*********************100%***********************]  1 of 1 completed


Price        Date               Close    Volume
Ticker                           GOOG      GOOG
0      2024-08-01  172.02812194824219  17177800
1      2024-08-02  167.98803710937500  18907800
2      2024-08-05  160.24702453613281  34907800
3      2024-08-06  160.14724731445312  36146500
4      2024-08-07  160.35675048828125  19334200
5      2024-08-08  163.43917846679688  15733600
6      2024-08-09  164.98539733886719  13549200
7      2024-08-12  163.54891967773438  12435000
8      2024-08-13  165.52406311035156  12717600
9      2024-08-14  161.63362121582031  22515900
10     2024-08-15  162.77082824707031  18392500
11     2024-08-16  164.33699035644531  16853100
12     2024-08-19  167.98803710937500  13100800
13     2024-08-20  168.54667663574219  12622500
14     2024-08-21  167.21992492675781  15269600
15     2024-08-22  165.08515930175781  19123800
16     2024-08-23  167.02040100097656  14281600
17     2024-08-26  167.51918029785156  11990300
18     2024-08-27  165.97299194335938  13718200
19     2024-08-28  164.09756469726562  15208700
20     2024-08-29  163.00025939941406  17133800
21     2024-08-30  164.70608520507812  18498800

####**Índice de Precio**



\begin{align}
         I_t^{(P)} = {p_t\over p_0}*100
\end{align}


In [ ]:
data=GOOG
col1=1
fila0=0
valores=[]
for i in data.iloc[:,col1]:
  valores.append((i/data.iloc[fila0, col1])*100)
valores

[100.0,
 97.65149744523589,
 93.15164446447079,
 93.09364393493546,
 93.21542819407604,
 95.00724452248019,
 95.90606202659474,
 95.0710371220242,
 96.21918860461228,
 93.95767353924305,
 94.6187323349626,
 95.5291428490332,
 97.65149744523589,
 97.97623477308701,
 97.2049935981217,
 95.96405368619132,
 97.08901027892853,
 97.37895083703395,
 96.48015107279691,
 95.38996463998915,
 94.75210073412048,
 95.74369779763856]

Se verifica este resultado para los dos primeros casos del siguiente modo:

$I_0^{(P)} = {p_0\over p_0}*100= {172.4499969482422\over 172.4499969482422}*100 = 100.0$

$I_1^{(P)} = {p_1\over p_0}*100= {168.39999389648438\over 172.4499969482422}*100 = 97.6514913752226$

Y así sucesivamente

####**Índice de Cantidad**



\begin{align}
         I_t^{(q)} = {q_t\over q_0}*100
\end{align}

In [ ]:
data=GOOG
col1= 2
fila0=0

valores=[]
for i in data.iloc[:,col1]:
  valores.append((i/data.iloc[fila0, col1])*100)
valores

[100.0,
 110.07113832970461,
 203.2146142113658,
 210.4256656847792,
 112.5534119619509,
 91.59263700823156,
 78.87622396348776,
 72.38994516177858,
 74.03509180453841,
 131.07557428774348,
 107.07133625959086,
 98.10976958632654,
 76.2658780519042,
 73.48147026976679,
 88.89147620766337,
 111.32857525410704,
 83.1398665719708,
 69.80113867899266,
 79.8600519274878,
 88.53694885258881,
 99.74385544132544,
 107.69015822747964]

Se verifica este resultado del siguiente modo:

$I_0^{(q)} = {q_0\over q_0}*100= {17177800\over 17177800}*100=100.0$

$I_1^{(q)} = {q_1\over q_0}*100= {18907800\over 17177800}*100=110.07113832970461$

Y así sucesivamente

####**Índice de Precio Agregado Simple**



\begin{align}
         I_t^{(Ag.Sim)} = {\sum_{k=1}^n p_{k;t}\over \sum_{k=1}^n p_{k;0}}*100
\end{align}


In [ ]:
pd.set_option("display.precision", 14) #visualizar cantidad de decimales completa
data=pd.DataFrame(yf.download(['GOOG', 'AMZN'], start='2024-08-01', end='2024-09-01')['Close']).reset_index()
data

[*********************100%***********************]  2 of 2 completed


Ticker       Date                AMZN                GOOG
0      2024-08-01  184.07000732421875  172.02812194824219
1      2024-08-02  167.89999389648438  167.98803710937500
2      2024-08-05  161.02000427246094  160.24702453613281
3      2024-08-06  161.92999267578125  160.14724731445312
4      2024-08-07  162.77000427246094  160.35675048828125
5      2024-08-08  165.80000305175781  163.43917846679688
6      2024-08-09  166.94000244140625  164.98539733886719
7      2024-08-12  166.80000305175781  163.54891967773438
8      2024-08-13  170.22999572753906  165.52406311035156
9      2024-08-14  170.10000610351562  161.63362121582031
10     2024-08-15  177.58999633789062  162.77082824707031
11     2024-08-16  177.05999755859375  164.33699035644531
12     2024-08-19  178.22000122070312  167.98803710937500
13     2024-08-20  178.88000488281250  168.54667663574219
14     2024-08-21  180.11000061035156  167.21992492675781
15     2024-08-22  176.13000488281250  165.08515930175781
16     2024-08-23  177.03999328613281  167.02040100097656
17     2024-08-26  175.50000000000000  167.51918029785156
18     2024-08-27  173.11999511718750  165.97299194335938
19     2024-08-28  170.80000305175781  164.09756469726562
20     2024-08-29  172.11999511718750  163.00025939941406
21     2024-08-30  178.50000000000000  164.70608520507812

In [ ]:
data= data
cols= [1,2]
fila1= 21
filabase1=0

valornum=[]
for i in cols:
  for j in data.iloc[fila1:fila1+1,i]:
    valornum.append(j)
numerador=sum(valornum)
valordenom=[]
for i in cols:
  for j in data.iloc[filabase1:filabase1+1,i]:
    valordenom.append(j)
denominador=sum(valordenom)
IPAS=(numerador/denominador)*100
IPAS

96.37963723827407

Se verifica este resultado del siguiente modo:

$I_{21}^{(Ag.Sim)} = {{\sum_{k=1}^2 p_{k;t}\over \sum_{k=1}^2 p_{k;0}}*100 }= {{178.5 + 165.11000061035156 \over 184.07000732421875 + 172.4499969482422} *100 = 96.37888379126035}$

####**Índice de Precio Agregado Ponderado**



\begin{align}
         I_t^{(Ag.Pond)} = {\sum_{k=1}^n q_k*p_{k;t}\over \sum_{k=1}^n q_k*p_{k;0}}*100
\end{align}

In [ ]:
data= data
cols= [1,2]
cants= [20,30]
fila1= 21
filabase1=0

valornum=[]
for i,k in zip(cols,cants):
  for j in data.iloc[fila1:fila1+1,i]:
    valornum.append(j*k)
numerador=sum(valornum)
valordenom=[]
for i,k in zip(cols, cants):
  for j in data.iloc[filabase1:filabase1+1,i]:
    valordenom.append(j*k)
denominador=sum(valordenom)
IPAP=(numerador/denominador)*100
IPAP

96.25591358841913

####**Índice de Laspeyres**



\begin{align}
         IL_t = {\sum_{k=1}^n q_{i,0}*p_{i;t}\over \sum_{k=1}^n q_{i,0}*p_{i;0}}*100
\end{align}

In [ ]:
pd.set_option("display.precision", 14) #visualizar cantidad de decimales completa
dataall=pd.DataFrame(yf.download(['GOOG', 'AMZN'], start='2024-08-01', end='2024-09-01')[['Close', 'Volume']]).reset_index()
dataall

[*********************100%***********************]  2 of 2 completed


Price        Date               Close                         Volume          
Ticker                           AMZN                GOOG       AMZN      GOOG
0      2024-08-01  184.07000732421875  172.02812194824219   70435600  17177800
1      2024-08-02  167.89999389648438  167.98803710937500  141448400  18907800
2      2024-08-05  161.02000427246094  160.24702453613281   83149400  34907800
3      2024-08-06  161.92999267578125  160.14724731445312   59950800  36146500
4      2024-08-07  162.77000427246094  160.35675048828125   48408200  19334200
5      2024-08-08  165.80000305175781  163.43917846679688   44616200  15733600
6      2024-08-09  166.94000244140625  164.98539733886719   36401000  13549200
7      2024-08-12  166.80000305175781  163.54891967773438   30072800  12435000
8      2024-08-13  170.22999572753906  165.52406311035156   39237900  12717600
9      2024-08-14  170.10000610351562  161.63362121582031   28843800  22515900
10     2024-08-15  177.58999633789062  162.77082824707031   51698500  18392500
11     2024-08-16  177.05999755859375  164.33699035644531   31489200  16853100
12     2024-08-19  178.22000122070312  167.98803710937500   31129800  13100800
13     2024-08-20  178.88000488281250  168.54667663574219   26255200  12622500
14     2024-08-21  180.11000061035156  167.21992492675781   35599100  15269600
15     2024-08-22  176.13000488281250  165.08515930175781   32047500  19123800
16     2024-08-23  177.03999328613281  167.02040100097656   29150100  14281600
17     2024-08-26  175.50000000000000  167.51918029785156   22366200  11990300
18     2024-08-27  173.11999511718750  165.97299194335938   29842000  13718200
19     2024-08-28  170.80000305175781  164.09756469726562   29045000  15208700
20     2024-08-29  172.11999511718750  163.00025939941406   26407800  17133800
21     2024-08-30  178.50000000000000  164.70608520507812   43429400  18498800

In [ ]:
data= dataall
colsp= [1,2]
colsq= [3,4]
fila1= 21
filabase1=0

valornum=[]
for i,k in zip(colsp, colsq):
  for j in data.iloc[fila1:fila1+1,i]:
    valornum.append(j*data.iloc[filabase1:filabase1+1,k])
numerador=sum(valornum)
valordenom=[]
for i,k in zip(colsp, colsq):
  for j in data.iloc[filabase1:filabase1+1,i]:
    valordenom.append(j*data.iloc[filabase1:filabase1+1,k])
denominador=sum(valordenom)
IL=list((numerador/denominador)*100)
IL[0]

96.74561220724019

####**Índice de Paschee**



\begin{align}
         IP_t = {\sum_{k=1}^n q_{i,t}*p_{i;t}\over \sum_{k=1}^n q_{i,t}*p_{i;0}}*100
\end{align}

In [ ]:
data= dataall
colsp= [1,2]
colsq= [3,4]
fila1= 21
filabase1=0

valornum=[]
for i,k in zip(colsp, colsq):
  for j in data.iloc[fila1:fila1+1,i]:
    valornum.append(j*data.iloc[fila1:fila1+1,k])
numerador=sum(valornum)
valordenom=[]
for i,k in zip(colsp, colsq):
  for j in data.iloc[filabase1:filabase1+1,i]:
    valordenom.append(j*data.iloc[fila1:fila1+1,k])
denominador=sum(valordenom)
IP=list((numerador/denominador)*100)
IP[0]

96.62366959238139

####**Índice de Fisher**



\begin{align}
         IF_t = \sqrt {IL_t * IP_t}
\end{align}

In [ ]:
data= dataall
colsp= [1,2]
colsq= [3,4]
fila1= 21
filabase1=0

valornum=[]
for i,k in zip(colsp, colsq):
  for j in data.iloc[fila1:fila1+1,i]:
    valornum.append(j*data.iloc[filabase1:filabase1+1,k])
numerador=sum(valornum)
valordenom=[]
for i,k in zip(colsp, colsq):
  for j in data.iloc[filabase1:filabase1+1,i]:
    valordenom.append(j*data.iloc[filabase1:filabase1+1,k])
denominador=sum(valordenom)
IL=list((numerador/denominador)*100)

valornum1=[]
for i,k in zip(colsp, colsq):
  for j in data.iloc[fila1:fila1+1,i]:
    valornum1.append(j*data.iloc[fila1:fila1+1,k])
numerador1=sum(valornum1)
valordenom1=[]
for i,k in zip(colsp, colsq):
  for j in data.iloc[filabase1:filabase1+1,i]:
    valordenom1.append(j*data.iloc[fila1:fila1+1,k])
denominador1=sum(valordenom1)
IP=list((numerador1/denominador1)*100)

IF=(IL[0]*IP[0])**0.5
IF


96.68462167493357

Se verifica este resultado del siguiente modo:

Dados $IL_{21}=96.74561220724019$ y $IP_{21}=96.62366959238139$

$IF_{21} = \sqrt {IL_{21} * IP_{21}} = \sqrt {96.74561220724019 * 96.62366959238139}= 96.68462167493357$

####**Cambio de base**



\begin{align}
         I_{baseY, t} = {I_{baseX, t}\over I_{baseX, Y}} *100
\end{align}

In [ ]:
pd.set_option("display.precision", 14) #visualizar cantidad de decimales completa
data=pd.DataFrame(yf.download(['GOOG', 'AMZN'], start='2024-08-01', end='2024-09-01')['Close']).reset_index()
data

[*********************100%***********************]  2 of 2 completed


Ticker       Date                AMZN                GOOG
0      2024-08-01  184.07000732421875  172.02812194824219
1      2024-08-02  167.89999389648438  167.98803710937500
2      2024-08-05  161.02000427246094  160.24702453613281
3      2024-08-06  161.92999267578125  160.14724731445312
4      2024-08-07  162.77000427246094  160.35675048828125
5      2024-08-08  165.80000305175781  163.43917846679688
6      2024-08-09  166.94000244140625  164.98539733886719
7      2024-08-12  166.80000305175781  163.54891967773438
8      2024-08-13  170.22999572753906  165.52406311035156
9      2024-08-14  170.10000610351562  161.63362121582031
10     2024-08-15  177.58999633789062  162.77082824707031
11     2024-08-16  177.05999755859375  164.33699035644531
12     2024-08-19  178.22000122070312  167.98803710937500
13     2024-08-20  178.88000488281250  168.54667663574219
14     2024-08-21  180.11000061035156  167.21992492675781
15     2024-08-22  176.13000488281250  165.08515930175781
16     2024-08-23  177.03999328613281  167.02040100097656
17     2024-08-26  175.50000000000000  167.51918029785156
18     2024-08-27  173.11999511718750  165.97299194335938
19     2024-08-28  170.80000305175781  164.09756469726562
20     2024-08-29  172.11999511718750  163.00025939941406
21     2024-08-30  178.50000000000000  164.70608520507812

Se crea un nuevo DataFrame directamente con los indice de precio (utilizando la función IPrecio creada previamente) de cada activo. En el casdo de AMZN se toma como período base 1/8/2024 y en el caso de GOOG 23/8/2024

In [ ]:
#Para Amazon
data=data
col1=1
fila0=0
valoresA=[]
for i in data.iloc[:,col1]:
  valoresA.append((i/data.iloc[fila0, col1])*100)
valoresA

[100.0,
 91.21529158237458,
 87.47758888760308,
 87.97195970691719,
 88.42831411733458,
 90.07442628049644,
 90.6937555271349,
 90.61769784034301,
 92.48111531157704,
 92.41049564576997,
 96.47959432363453,
 96.19166106008915,
 96.82185805903104,
 97.18041927804965,
 97.84864097555443,
 95.68642248847155,
 96.18079330778568,
 95.34415875307505,
 94.05116978794702,
 92.79078407972935,
 93.50789822810044,
 96.97397343261478]

In [ ]:
#Para google
data=data
col1=2
fila0=16
valoresG=[]
for i in data.iloc[:,col1]:
  valoresG.append((i/data.iloc[fila0, col1])*100)
valoresG

[102.99826902417529,
 100.57935204477972,
 95.94458136595891,
 95.88484182451266,
 96.01027750337137,
 97.85581730572018,
 98.78158377664447,
 97.92152257901603,
 99.10409873185716,
 96.77477736080591,
 97.4556564776291,
 98.39336354813591,
 100.57935204477972,
 100.91382587134173,
 100.11946081112575,
 98.84131418220734,
 100.0,
 100.29863375604762,
 99.3728855568901,
 98.2500124019617,
 97.59302362018698,
 98.61435143130515]

In [ ]:
pd.set_option("display.precision", 14)
INDICES=pd.DataFrame({'IP_AMZN': valoresA, 'IP_GOOG': valoresG})
INDICES

IP_AMZN             IP_GOOG
0   100.00000000000000  102.99826902417529
1    91.21529158237458  100.57935204477972
2    87.47758888760308   95.94458136595891
3    87.97195970691719   95.88484182451266
4    88.42831411733458   96.01027750337137
5    90.07442628049644   97.85581730572018
6    90.69375552713490   98.78158377664447
7    90.61769784034301   97.92152257901603
8    92.48111531157704   99.10409873185716
9    92.41049564576997   96.77477736080591
10   96.47959432363453   97.45565647762911
11   96.19166106008915   98.39336354813591
12   96.82185805903104  100.57935204477972
13   97.18041927804965  100.91382587134173
14   97.84864097555443  100.11946081112575
15   95.68642248847155   98.84131418220734
16   96.18079330778568  100.00000000000000
17   95.34415875307505  100.29863375604762
18   94.05116978794702   99.37288555689010
19   92.79078407972935   98.25001240196170
20   93.50789822810044   97.59302362018698
21   96.97397343261478   98.61435143130515

Como los índices tienen período base diferentes, para que sean comparables, se realiza el cambio de base para AMZN, tomando como nuevo período base 23/8/2024

Se crea una función para recalcular un índice cambiando la base

In [ ]:
data= INDICES
col1= 0
filabasenew=16

valores=[]
for i in data.iloc[:,col1]:
  valores.append((i/data.iloc[filabasenew, col1])*100)

INDICES['IP_AMZN_NB']=valores
INDICES

IP_AMZN             IP_GOOG          IP_AMZN_NB
0   100.00000000000000  102.99826902417529  103.97086212420039
1    91.21529158237458  100.57935204477972   94.83732504729804
2    87.47758888760308   95.94458136595891   90.95120333190462
3    87.97195970691719   95.88484182451266   91.46520493483598
4    88.42831411733458   96.01027750337137   91.93968054968876
5    90.07442628049644   97.85581730572018   93.65115755725947
6    90.69375552713490   98.78158377664447   94.29507951437679
7    90.61769784034301   97.92152257901603   94.21600168170755
8    92.48111531157704   99.10409873185716   96.15341289152254
9    92.41049564576997   96.77477736080591   96.07998901615370
10   96.47959432363453   97.45565647762911  100.31066599221393
11   96.19166106008915   98.39336354813591  100.01129929576345
12   96.82185805903104  100.57935204477972  100.66652054864417
13   97.18041927804965  100.91382587134173  101.03931973930086
14   97.84864097555443  100.11946081112575  101.73407559909755
15   95.68642248847155   98.84131418220734   99.48599839706863
16   96.18079330778568  100.00000000000000  100.00000000000000
17   95.34415875307505  100.29863375604762   99.13014384063840
18   94.05116978794702   99.37288555689010   97.78581206642400
19   92.79078407972935   98.25001240196170   96.47537817949988
20   93.50789822810044   97.59302362018698   97.22096794197593
21   96.97397343261478   98.61435143130515  100.82467621398263

Se verifica este resultado para algunos casos del siguiente modo:

$I_{16, 16} = {I_{16, 16}\over I_{16, 16}} *100 = {96.18079330778568\over 96.18079330778568} *100 = 100.0$

$I_{16, 0} = {I_{16, 0}\over I_{16, 16}} *100 = {100.0\over 96.18079330778568} *100 = 103.97086212420039$

$I_{16, 21} = {I_{16, 21}\over I_{16, 16}} *100 = {96.97397343261478\over 96.18079330778568} *100 = 100.82467621398263$